src/java/org/apache/cassandra/io/sstable/SSTable.java

```diff
@@ -391,6 +404,47 @@ public synchronized void addComponents(Collection<Component> newComponents)
         components.addAll(componentsToAdd);
     }
 
+    /**
+     * Registers new custom components into sstable and update size tracking
+     * @param newComponents collection of components to be added
+     * @param tracker used to update on-disk size metrics
+     */
+    public synchronized void registerComponents(Collection<Component> newComponents, Tracker tracker)
+    {
+        Collection<Component> componentsToAdd = new HashSet<>(Collections2.filter(newComponents, x -> !components.contains(x)));
+        appendTOC(descriptor, componentsToAdd);
+        components.addAll(componentsToAdd);
+
+        if (tracker == null)
```

Comment:
I'm trying to find a way `tracker` could be null. `StorageAttachedIndexBuilder` gets it from the CFS, but that doesn't look like it should ever be null...

Reply:
I can't see how it can be `null` either. It's created in the table constructor and is `final`. I'll remove this check.